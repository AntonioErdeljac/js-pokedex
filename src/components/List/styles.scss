@import '../../scss/bootstrap';

.pd-list {
  margin: map-get($map: $spacers, $key: 5) * 3 0px;

  @media only screen and (max-width: map-get($map: $grid-breakpoints, $key: lg)) {
    margin: map-get($map: $spacers, $key: 5) * 2 0px !important;
  }

  &__filters {
    margin-bottom: map-get($map: $spacers, $key: 5);

    @media only screen and (max-width: map-get($map: $grid-breakpoints, $key: lg)) {
      margin-bottom: map-get($map: $spacers, $key: 3);
    }
  }

  &__sort {
    @media only screen and (max-width: map-get($map: $grid-breakpoints, $key: lg)) {
      margin-top: map-get($map: $spacers, $key: 3);
    }
  }
}
