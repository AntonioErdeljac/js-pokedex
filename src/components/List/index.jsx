import cn from 'classnames/bind';
import PropTypes from 'prop-types';
import React from 'react';
import Visibility from 'react-visibility-sensor';
import { Container, Form, Row } from 'react-bootstrap';
import { isEmpty } from 'lodash';

import empty from './images/empty.png';
import styles from './styles.scss';
import { Item, BottomPixel, Loader } from './components';

import FadeIn from '../FadeIn';
import Info from '../Info';

const cx = cn.bind(styles);

const List = ({ items, onScroll, isLoading, hasLoaded, onSearch, searchValue }) => {
  let content = (
    <>
      {onSearch ? (
        <Form.Control
          value={searchValue}
          onChange={onSearch}
          className={cx('pd-list__input')}
          placeholder="Search"
        />
      ) : null}
      <Row className="mt-3">
        {items.map((item, index) => (
          <Item key={item.name} photo={item.photo} name={item.name} index={index} />
        ))}
        {isLoading && !isEmpty(items) ? (
          <Container>
            <Loader />
          </Container>
        ) : null}
      </Row>
      {onScroll ? (
        <Visibility onChange={onScroll}>
          <BottomPixel />
        </Visibility>
      ) : null}
    </>
  );

  if (hasLoaded && isEmpty(items)) {
    content = (
      <>
        {onSearch ? (
          <Form.Control
            value={searchValue}
            onChange={onSearch}
            className={cx('pd-list__input')}
            placeholder="Search"
          />
        ) : null}
        <Info text="No items found." image={empty} />
      </>
    );
  }

  if (isLoading && isEmpty(items)) {
    content = <Loader />;
  }

  return (
    <FadeIn>
      <Container className={cx('pd-list')}>{content}</Container>
    </FadeIn>
  );
};

List.defaultProps = {
  onScroll: null,
  isLoading: false,
  hasLoaded: false,
  onSearch: null,
  searchValue: '',
};

List.propTypes = {
  isLoading: PropTypes.bool,
  hasLoaded: PropTypes.bool,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      src: PropTypes.string,
    }),
  ).isRequired,
  onScroll: PropTypes.func,
  onSearch: PropTypes.func,
  searchValue: PropTypes.string,
};

export default List;
