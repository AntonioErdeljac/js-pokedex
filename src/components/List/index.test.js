import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';

import List from './index';

const EXPECTED_ITEMS_LENGTH = 2;

const makeProps = ({ items, hasLoaded, isLoading } = {}) => ({
  items: items || [{ name: 'Lorem' }, { name: 'Ipsum' }],
  hasLoaded: hasLoaded || false,
  isLoading: isLoading || false,
});

const renderComponent = (props) => {
  return shallow(<List {...props} />);
};

describe('List', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = renderComponent(makeProps());
  });

  it('renders correctly according to snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correct number of items', () => {
    expect(wrapper.find('Memo(Item)').length).toEqual(EXPECTED_ITEMS_LENGTH);
  });
});
