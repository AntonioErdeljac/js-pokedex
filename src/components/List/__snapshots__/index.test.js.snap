// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly according to snapshot 1`] = `
<Memo(FadeIn)>
  <Container
    className="pd-list"
    fluid={false}
  >
    <Row
      className="pd-list__filters"
      noGutters={false}
    >
      <Col
        lg={10}
        xs={12}
      >
        <FormControl
          onChange={null}
          placeholder="Search"
          value=""
        />
      </Col>
      <Col
        lg={2}
        xs={12}
      >
        <Dropdown
          navbar={false}
          onSelect={null}
        >
          <DropdownToggle
            className="pd-list__sort w-100"
            variant="info"
          >
            Sort By
          </DropdownToggle>
          <DropdownMenu
            align="left"
            alignRight={false}
            className="w-100"
            flip={true}
          >
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              eventKey={null}
            >
              Default
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              eventKey="asc"
            >
              Name (A-Z)
            </DropdownItem>
            <DropdownItem
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              disabled={false}
              eventKey="desc"
            >
              Name (Z-A)
            </DropdownItem>
          </DropdownMenu>
        </Dropdown>
      </Col>
    </Row>
    <Row
      className="mt-3"
      noGutters={false}
    >
      <Memo(Item)
        index={0}
        key="Lorem"
        name="Lorem"
      />
      <Memo(Item)
        index={1}
        key="Ipsum"
        name="Ipsum"
      />
    </Row>
  </Container>
</Memo(FadeIn)>
`;
